### DISCUSSION TITLE
Handling duplicate points without throwing an error

### DISCUSSION THREAD

**Discussion by Kevin-Mattheus-Moerman** [DanielVandH/DelaunayTriangulation.jl#215] [Ideas]
@DanielVandH I have a question about duplicate point errors. Could DelaunayTriangulation handle the duplicate points (e.g. by removing the duplicates?) and issue a warning rather than an error? 

```
ERROR: DuplicatePointsError: There were duplicate points. The following points are duplicated:
    (0.625, -1.55) at indices [4, 25].
```

**Comment by DanielVandH** [comment]
In general I prefer to make the user make the decision here. But maybe this would be nice. The identified duplicates could be passed to the existing kwarg skip_points (merged with the existing if needing). Do you want to make a pr that does that?

**Comment by EdsterG** [comment]
I generally agree with your philosophy @DanielVandH, however there is one use case where making the user deal with it is quite annoying. Specifically the Makie case, there have been many times where I wanted to visualize the data without care for duplicates. It would be super convenient to allow Makie to filter those points. With that said, Makie could in theory pipe all of the data through a unique function call, but that generally seems wasteful because most data won't have duplicates. If this library already has some logic for duplicate detection, maybe it can silently skip those points (put behind a flag). Then Makie can set that flag accordingly?

**Comment by DanielVandH** [comment]
I think changing the default behaviour to skipping duplicates and displaying a warning to the user is sufficient. Making this a flag that Makie sets would be annoying I think since I don't think I want to make this a new keyword argument to `triangulate` (I could make this something I do in v2 of the package where I make these changes a lot less annoying - a v2 branch exists but haven't gotten to it in a while), and the alternative is making a flag similar to https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/setup.jl#L42 that Makie sets and resets which might be a bit awkward. Does the warning need to be possible to silence? 

My suggestion in my last reply would be pretty simple to implement I think. 

https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/algorithms/triangulation/main.jl#L154

would be changed to `check_arguments && (skip_points = check_args(points, boundary_nodes, full_polygon_hierarchy, boundary_curves, skip_points)`, and the code here https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/main/src/algorithms/triangulation/check_args.jl would be modified accordingly, e.g. https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/algorithms/triangulation/check_args.jl#L26 becomes `skip_points = has_unique_points(points, skip_points)` and you make use of https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/algorithms/triangulation/check_args.jl#L57 to figure out what points to merge into the existing `skip_points` (`skip_points` can probably be converted into a `Set` so that mutation works regardless of what the user provides, be it a `Tuple` or a `Vector`). Here https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/algorithms/triangulation/check_args.jl#L121 you would warn instead of `throw`.

**Reply by EdsterG** [reply]
> Does the warning need to be possible to silence?

I would vote yes, because inline warnings in a notebook can be quite annoying, since they mess with the formatting of the output cell.

**Reply by EdsterG** [reply]
Out of curiosity, for non-plotting use cases why is it important to warn/error on duplicate points?

**Reply by DanielVandH** [reply]
Ah okay, I don't use notebooks often so I'm not familiar. I'm happy to have a flag similar to the `INF_WARN` I linked then. Makie wouldn't be the one to set this though, that would be the user. See 

https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/algorithms/voronoi/unbounded.jl#L38

https://github.com/JuliaGeometry/DelaunayTriangulation.jl/blob/9134c8ebba348338dcf394bdf615843caf7f0bac/src/setup.jl#L42

**Reply by DanielVandH** [reply]
> Out of curiosity, for non-plotting use cases why is it important to warn/error on duplicate points?

At least in my use cases, I am typically always interested in knowing when I have duplicate points since that indicates something wrong is occurring. e.g. for cell biology if I have two-overlapping cells then something's gone wrong in the simulation and any future simulations would probably be nonsense. 

I also just like the philosophy of forcing the user to make a decision, though I recognise that can be a bit annoying especially now that Makie uses the package - ideally users just trying to plot shouldn't even really have to be aware that this package exists.

**Comment by DanielVandH** [comment]
v1.6.4 will now have a fix for this @Kevin-Mattheus-Moerman  @EdsterG .

**Reply by EdsterG** [reply]
Yeah, this is great. Thanks for the fix!
